### ISSUE TITLE
People: Facial Recognition

### ISSUE THREAD

**Issue by lastzero** [photoprism/photoprism#22]
[<img alt="Issuehunt badges" src="https://img.shields.io/badge/IssueHunt-%24900%20Rewarded-%237E24E3.svg" />](https://issuehunt.io/r/photoprism/photoprism/issues/22)

**As a user, I want to tag familiar human faces so that I can find all photos of these persons.**

PhotoPrism must be able to detect and recognize faces. There are existing libraries for this, so we should try those first. See [Face Recognition](https://github.com/photoprism/photoprism/wiki/Face-Recognition) in our Wiki.

Here is a complete howto: https://hackernoon.com/face-recognition-with-go-676a555b8a7e

Acceptance Criteria:
- [x] You SHOULD use [go-face](https://github.com/Kagami/go-face) (unless you find a better library)
- [x] A unit test demonstrating face recognition with real images and faces MUST be delivered
- [x] A user interface MAY be developed and gives extra karma
- [x] You MUST update our Docker images in `/docker` as needed (go-face requires dlib as external dependency); let us know when we shall push new images to [Docker Hub](https://hub.docker.com/u/photoprism)
- [x] [Documentation](https://github.com/photoprism/photoprism/wiki/Face-Recognition) MUST be updated so that other developers know everything to continue working on this task

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $900.00)

- [<img src='https://avatars3.githubusercontent.com/u/44827199?v=4' alt='issuehunt' width=24 height=24> issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)
- [<img src='https://avatars2.githubusercontent.com/u/54749?v=4' alt='totempaaltj' width=24 height=24> totempaaltj](https://issuehunt.io/u/totempaaltj) ($30.00)
- [<img src='https://avatars.githubusercontent.com/u/64777347?v=4' alt='uluvil3gcs' width=24 height=24> uluvil3gcs](https://issuehunt.io/u/uluvil3gcs) ($22.00)
- [<img src='https://avatars3.githubusercontent.com/u/467291?v=4' alt='urbiops' width=24 height=24> urbiops](https://issuehunt.io/u/urbiops) ($20.00)
- [<img src='https://avatars1.githubusercontent.com/u/17299920?v=4' alt='luxumbra27' width=24 height=24> luxumbra27](https://issuehunt.io/u/luxumbra27) ($10.00)
- [<img src='https://avatars3.githubusercontent.com/u/67567419?v=4' alt='cloudfreeshop' width=24 height=24> cloudfreeshop](https://issuehunt.io/u/cloudfreeshop) ($50.00)
- [<img src='https://avatars0.githubusercontent.com/u/6235557?v=4' alt='kreusen' width=24 height=24> kreusen](https://issuehunt.io/u/kreusen) ($10.00)
- [<img src='https://avatars1.githubusercontent.com/u/16369490?v=4' alt='setime' width=24 height=24> setime](https://issuehunt.io/u/setime) ($4.00)
- [<img src='https://avatars.githubusercontent.com/u/627140?v=4' alt='spoorun' width=24 height=24> spoorun](https://issuehunt.io/u/spoorun) ($100.00)
- [<img src='https://avatars.githubusercontent.com/u/124109?v=4' alt='friesoft' width=24 height=24> friesoft](https://issuehunt.io/u/friesoft) ($10.00)
- [<img src='https://avatars.githubusercontent.com/u/17353007?v=4' alt='conrad82' width=24 height=24> conrad82](https://issuehunt.io/u/conrad82) ($10.00)
- [<img src='https://avatars.githubusercontent.com/u/4406403?v=4' alt='angelnu' width=24 height=24> angelnu](https://issuehunt.io/u/angelnu) ($20.00)
- [<img src='https://avatars.githubusercontent.com/u/19432439?v=4' alt='bitbacchus' width=24 height=24> bitbacchus](https://issuehunt.io/u/bitbacchus) ($5.00)
- [<img src='https://avatars.githubusercontent.com/u/15955392?v=4' alt='mauriceoegerli' width=24 height=24> mauriceoegerli](https://issuehunt.io/u/mauriceoegerli) ($50.00)
- [<img src='https://avatars.githubusercontent.com/u/49533775?v=4' alt='csev1755' width=24 height=24> csev1755](https://issuehunt.io/u/csev1755) ($5.00)
- [<img src='https://avatars.githubusercontent.com/u/10055792?v=4' alt='ayancey' width=24 height=24> ayancey](https://issuehunt.io/u/ayancey) ($5.00)
- [<img src='https://avatars.githubusercontent.com/u/6353339?v=4' alt='ark-' width=24 height=24> ark-](https://issuehunt.io/u/ark-) ($20.00)
- [<img src='https://avatars.githubusercontent.com/u/2584158?v=4' alt='mastert' width=24 height=24> mastert](https://issuehunt.io/u/mastert) ($5.00)
- $484.00 have been anonymously funded.
### Submitted pull Requests
- [#1541 People: Prepare for release #22](https://issuehunt.io/r/photoprism/photoprism/pull/1541)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/photoprism/photoprism/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

**Comment by husayt** [comment]
Here are some more libraries to consider:

- [Openface](https://cmusatyalab.github.io/openface/demo-2-comparison/)

- https://github.com/ageitgey/face_recognition

The latter is using Dlib state-of-the-art face recognition built with deep learning. The model has an accuracy of 99.38% on the Labeled Faces in the Wild benchmark

**Comment by lastzero** [comment]
I'd prefer a library implemented in Go to improve performance and reduce complexity.

**Comment by devzsolt** [comment]
As a proof of concept you can create a small interface to an external tool (eg. https://github.com/ageitgey/face_recognition using its CLI) and then later it's possible to replace it with a native Go lib. I think it would speed up the early development, this Python lib looks pretty much mature.

**Comment by Xeoncross** [comment]
The easiest way to do this is simply to slap https://github.com/esimov/pigo into the project as a pure Go include.

**Comment by lastzero** [comment]
@Xeoncross Thanks! But Pigo is just face detection, not face recognition...

See https://github.com/photoprism/photoprism/wiki/Face-Recognition

**Comment by lastzero** [comment]
This looks good to me on first sight: https://github.com/jdeng/goface

**Comment by trosel** [comment]
Could object detection be included in this request as well? Some libraries work with both faces and objects I think.

**Comment by lastzero** [comment]
@trosel We already have object detection / classification, without position however. Facial recognition is different from classification in that you need parameters for face parts as output to do the actual recognition based on a clusters (and not just detection).

What specific library do you refer to?

**Comment by cgudea** [comment]
I don't have any go experience but I'd really like to see this feature implemented. Is there anyone currently doing work on this? If so, would adding money onto the bounty help expedite the development?

**Comment by lastzero** [comment]
@cgudea You're most welcome to support our work via GitHub Sponsors:

https://github.com/sponsors/lastzero

No contributions there yet as GitHub Sponsors is pretty new. They will match any donation in the first year!

Don't expect to see a pull request for facial recognition anytime soon, it's too complex. We'll start working on it once albums and sharing are done. Financial support makes a major difference as we otherwise have to stop development frequently to do commercial projects.

**Comment by cgudea** [comment]
@lastzero totally understandable. I'm a huge fan of the goals of this project and will be looking to contribute after the holidays in January either financially or via development (I am a mobile developer at my dayjob).

**Comment by lastzero** [comment]
@cgudea Thank you <3

**Comment by ulinuxp** [comment]
@lastzero If you want to get rid of the dependency of dlib, you can try the way that ownphotos does facial clustering. https://github.com/hooram/ownphotos/blob/dev/api/face_clustering.py

You can use any tensorflow CNN model that extracts facial features, and do a kmeans on them with https://github.com/muesli/kmeans

**Comment by lastzero** [comment]
@ulinuxp Thank you! Hopefully we'll soon have time to start with this...

**Comment by guysoft** [comment]
> Here are some more libraries to consider:
> 
>     * [Openface](https://cmusatyalab.github.io/openface/demo-2-comparison/)
> 
>     * https://github.com/ageitgey/face_recognition
> 
> 
> The latter is using Dlib state-of-the-art face recognition built with deep learning. The model has an accuracy of 99.38% on the Labeled Faces in the Wild benchmark

The exact same engine is used here in go
https://github.com/Kagami/go-face

**Comment by dtaivpp** [comment]
Going to look into starting this. Also, I think there should be a UI. Don't get me wrong I think that grouping people under a random tag would be less useful than allowing users to name their people. 

Here are my thoughts for the logic and you all can let me know what you think. I haven' t looked too much over the code base but will do that before implementing this so it is integrated with native triggers. 

Backend: 
`
Application Starts
    if no models exist:
        start scanning all files and generating models
        models are randomly named and tags are added
    else:
        on new photo entry
        check against current models
        if no matching models
            create new random tag
        else
            update with random tag / or name if exists
`
Frontend: 
`
Show Sample image with peoples faces and tag
    allow the user to change the tag to name
    on name entry, update model to use name to tag
`

I just wanted to put some sample logic out there so we can break this down into a few steps and get it progressively added. I think the first phase would be getting Go-Face creating models and grouping faces by random tag.

**Comment by dtaivpp** [comment]
Going to look into starting this. Also, I think there should be a UI. Don't get me wrong I think that grouping people under a random tag would be less useful than allowing users to name their people. 

Here are my thoughts for the logic and you all can let me know what you think. I haven' t looked too much over the code base but will do that before implementing this so it is integrated with native triggers. 

Backend: 
`
Application Starts
    if no models exist:
        start scanning all files and generating models
        models are randomly named and tags are added
    else:
        on new photo entry
        check against current models
        if no matching models
            create new random tag
        else
            update with random tag / or name if exists
`
Frontend: 
`
Show Sample image with peoples faces and tag
    allow the user to change the tag to name
    on name entry, update model to use name to tag
`

I just wanted to put some sample logic out there so we can break this down into a few steps and get it progressively added. I think the first phase would be getting Go-Face creating models and grouping faces by random tag.

**Comment by dtaivpp** [comment]
Small update, it seems we will only need to tag the photos. I just noticed there is already support for re-naming tags. I'll start looking into just classifying and randomly naming them. Then users can simply update the labels. The tricky part will be photos with more than one person but we can cross that bridge when we get there.

**Comment by Gestas** [comment]
I've been testing the approach above vs requiring users to create a per-person. This is with go-face directly, no PhotoPrism integration at this point. I'm using my personal photos, ~18K images. Some with faces, others without. I'm getting a significantly better user experience with the latter approach. I'm not an expert in this space, perhaps I'm doing something wrong. 

When I use go-face to automatically create models I end up with collections of faces 100's big. For adults the rate of incorrect matches can be as high as 20%, for children 45%. The user experience required to correct the model is poor, I have to remove images. If instead I label 20-30 images first the run the classifier I get better results with a significantly better user experience. 

I suggest we prompt users to label a set of photos first then run the classifier. For images with unknown faces we could add an "unlabeled-face" tag that would allow users to easily find and tag more faces.

**Comment by dunderpate** [comment]
As a user, the approach of labeling a set of photos makes sense. Do we have a flow for re-labeling images that are incorrectly labeled, or to re-classify once the models are updated in order to correct mislabeled images? Thanks for the work on this, this project is getting really close to a viable Photos alternative once this is added, for my purposes at least.

**Comment by lastzero** [comment]
Labels are only meant for object classification, not for face recognition. We need to use a separate database table with different columns for this.

**Comment by dtaivpp** [comment]
@lastzero Can you explain to me why that is the case? I am new to this project and am not sure I understand the whole flow right now. Would adding names as labels interfere with other processes? 

@Gestas do you have a fork I could look at? I'm curious if the model may just need some fine-tuning. For example, we could put a higher threshold on what justifies a match and then as more data is added rerun and collect ones that previously didn't have a high enough confidence.

**Comment by guysoft** [comment]
You will likely need a "face id" and then that face id would link to
information about the person (for example a name). That would require
another table.

**Comment by Gestas** [comment]
@dtaivpp I think @lastzero is right, we need to be able to differentiate between labels (nouns) and faces during a search. E.g in English "Forest" can be a set of trees or a name for a human. A user might want to search for "Forest" the person and not a set of trees.

**Comment by Gestas** [comment]
Are there any changes to the search API in the works? I'd like to support compound face searches. Given the current design it looks like we would add a "face" parameter to the API but that would only support searching on a single name.

**Comment by lastzero** [comment]
We could split the search string after `, `

**Comment by dtaivpp** [comment]
@lastzero that is probably the best option with the least time to develop. I could see the benefit of being able to search for pictures with multiple people's faces @Gestas 

Another potential challenge is identifying which face the tag is being related to. Say for example every time I go out with my friend Mike, I also go with Carl. I will only have pictures of both Mike and Carl so on the front end there needs to be a way to distinguish which is which. That way when I do take a picture of just me and Mike it isn't confused.

**Comment by Gestas** [comment]
@lastzero I think it would make sense to extend the ```,``` model to all parameters. E,g - ```&camera=2,3```, ```&year=2018,2019```. Should I open a separate feature request for this?

@dtaivpp that's doable. I think it's not something we need to include in the MVP release, do you?

**Comment by ThomasEg** [comment]
> @cgudea You're most welcome to support our work via GitHub Sponsors:
> 
> https://github.com/sponsors/lastzero
> 
> No contributions there yet as GitHub Sponsors is pretty new. They will match any donation in the first year!
> 
> Don't expect to see a pull request for facial recognition anytime soon, it's too complex. We'll start working on it once albums and sharing are done. Financial support makes a major difference as we otherwise have to stop development frequently to do commercial projects.

Done:-) I’ve just sponsered you, hoping to bump the face recognition a few notches up;-)

**Comment by totti4ever** [comment]
One question: Would the recognized face just be stored in an internal database of photoprism or would you consider an option to also store it as label to the picture itself? This should only happen on specific request of the user to not update thousands of files coincidentally, but on the other hand I don't want a lock-in and be able to access the recognized people by using any other software that reads standard meta tags

**Comment by lastzero** [comment]
@totti4ever We're currently not modifying ANY originals to protect them. But we can certainly provide metadata incl faces and labels as sidecar file, database dump or via API.

**Comment by totti4ever** [comment]
I totally understand that you don't want to edit files being a front-end software. Although I miss software to properly process my images and add information about recognized people in the pictures.

So if you do that anyway, I'd just love to also write them to the files - whether through the UI or by some CLI or API call, I wouldn't mind

**Comment by eliliam** [comment]
@lastzero Are there any updates on this? With all of these donations coming in, there hasn't been any updates here in over six months. Seriously, if PhotoPrism pulls this off, that would be a huge win for the project. It seems as if all other open source solutions have dropped off the map in terms of facial recognition.

**Comment by guysoft** [comment]
@lastzero I've been doing computer-visiom detection the past few months at work. If I get today a poof of concept function in go, would you consider an initial UI and backend?

**Comment by lastzero** [comment]
@guysoft Any help with this much appreciated! We should tag a stable release first. Batch edit & facial recognition is on the top of our todo list for the time after.

**Comment by spiritbro1** [comment]
when do you think the stable release be released? i want to work on this one looks interesting for me too

**Comment by elboletaire** [comment]
Dude this rises like bitcoin in early 2016. @guysoft have you started working on it? I've been subscribed to this issue for a while, and now looks like it's worth taking some time for it :joy: 

I'm really busy right now but, if no one starts doing this, I may try to do it during the next weeks.

Ofc after knowing no one is currently working on it. And I don't want to add any pressure. If someone is working on it, please, update us! :)

**Comment by guysoft** [comment]
I was searching for face clustering tools for go and got stuck.
there is detection and recognition, but no clustering

**Comment by Extarys** [comment]
@elboletaire [Off topic:] the goal is to wait as long as possible and get all the gold at the end. It looks like that though

**Comment by jimmcslim** [comment]
I'm wondering if this could be implemented with some sort of provider pattern? Ship a default local implementation, but also provide scope for an implementation that relies upon a cloud-based facial recognition provider, such as the Computer Vision APIs that are part of Azure Cognitive Services? I appreciate this might not be in the spirit of the 'own your data' ethos of PhotoPrism, but just putting it out there as a suggestion.

**Comment by guysoft** [comment]
I would vote against that because you want it ti be self-hosted. And it can be.
I seem to think at this point we could 
1. Implement something from scratch
2. Wait till something comes out in go
3. Port what ownphoto used and have a way to interface to it with go

**Comment by ChrisFab16** [comment]
> 
> 
> I would vote against that because you want it ti be self-hosted. And it can be.
> I seem to think at this point we could
> 
>     1. Implement something from scratch
> 
>     2. Wait till something comes out in go
> 
>     3. Port what ownphoto used and have a way to interface to it with go

As far as I can tell, there has been activity on the ownphoto fork in this space as well, so it's likely worth taking a look at the current state.

**Comment by mubashir-rehman** [comment]
Even though Picasa3D(by Google) is deprecated, it still works very well at recognizing faces with small errors. Also, Digikam is also working on such a feature and it looks promising(not in GO).

**Comment by guysoft** [comment]
I would advise against that since I think Photoprism stands for self hosting. The Readme says "Say goodbye to solutions that force you to upload your visual memories to the cloud". And using Picasa would be exactly that.
Moreover it should be possible to have a facial recognition core that is self-hosted like there is in ownphotos (which has been abandoned).

**Comment by guysoft** [comment]
I think it might be worth just using this:
http://openbiometrics.org/

**Comment by mubashir-rehman** [comment]
@guysoft Picasa3D is also deprecated and works offline and there is no working option to upload files to cloud(even if available I never used it).
Sorry for cluttering this issue.

**Comment by crsawyer314** [comment]
I'm brand new to the project, but I'd like to tackle this issue. If anyone else is working on it already, please lmk, but otherwise could I get it assigned to me?

**Comment by lastzero** [comment]
@crsawyer314 Thank you! We know facial recognition is pretty popular (and also want it for ourselves), but need to get a stable release and geodata api out of the door first. As there are various ways to solve this (and many contributors who might be able to help) it should be a coordinated effort.

In a nutshell: 

- We certainly don't want to use external services or scripting languages
- Second best option is using a popular library like [OpenCV](https://opencv.org/)
- The ideal solution is [Go-based](https://github.com/esimov/pigo) and doesn't require any external dependencies

What about we're starting with this in January? Of course, prototypes and links to libraries may be posted at any time.

**Comment by guysoft** [comment]
FYI pigo only detects location of faces, it can't compute distance between them (as in so you can cluster them and then understand they are the same persion). See https://github.com/esimov/pigo/issues/38

**Comment by lastzero** [comment]
Would be great to know the roadmap of related Go machine learning projects / libraries. I know recognition was not possible when I checked last time, just detection. Maybe there is something in the pipeline it's worth waiting for?

**Comment by mogwai** [comment]
I'm not a go programmer but I can create a PR for this that uses https://github.com/orktes/go-torch and a state of the art face recognition model and detection model?

**Comment by dtaivpp** [comment]
@mogwai the biggest issue is not recognizing but doing clustering to detect similar faces.

**Comment by mogwai** [comment]
Yes so the model, given a cropped RGB face would create a vector (or set of face features) which could be compared to another face vector. The "distance" between these two vectors would be close to 0 if they are similar. You could pre-compute these vectors when a photo is initially processed (or in batches on the gpu). Faces similar to each other will then need to be named by the user.

This is what is meant by face recognition. 

You'll need a separate model to do the face detection, which will find coordinates for bounding boxes in a given image.  This can be used to crop the images for the recognition (identification) model .

**Comment by denim2x** [comment]
I'll be looking into a ResNet-based solution and will share my results soon.

**Comment by MrPeteH** [comment]
I encourage looking at Picasa3 NOT to use it but to understand what is/was awesome about the UX. A short list:
- Per-face data tagged in the file itself
- Parameters include folders to scan/rescan/ignore for this operation, some kind of quality filter
- Basic operations include: Add name to face (new or existing), combine faces, ignore faces, manually add face rectangles on a photo, 
- Basic view modes include: group faces (ie show Joe(56) rather than 56 separate photos of Joe), individual, show whole photo (normally I just want to see thumbnails of recognized faces!)
- (Advanced feature: generate slide show with faces resized and recentered on eyes.)

Picasa3 FAILS:
- Database crash loses all ignored pix, often rebuilds face rectangles incorrectly and duplicated (perhaps the metadata is not stored correctly in photos?)

**Comment by nicrf** [comment]
Hi,

Being new to your project,In order to create a manual tag I want to know if you would be more in favor of a complete "entity" for the person "tag" or would you prefer to use "users" as the base person (name, Dob, etc.. )?

Because, I was tinking about that, and you don't want always create a new user for a peaple tag (like all friend you  know).

So if I am rigth , I will use two new entity, person  and person tag, where the person can be linked "Has One'  to user if it is  not null.

Something like that :
`
type Person struct {
	ID            	string    	`gorm:"type:VARBINARY(42);primary_key;auto_increment:false;" json:"PlaceID" yaml:"PlaceID"`
	PersonFullName    	string    	`gorm:"type:VARBINARY(755);unique_index;" json:"FullName" yaml:"FullName"`
	PersonUserId     	*string    	`gorm:"type:VARCHAR(42);" json:"UserId" yaml:"UserId,omitempty"`
	PersonBoD     	*time.Time 	`json:"BoD" yaml:"BoD,omitempty"`
	PersonDeadDate     	*time.Time 	`json:"DeadDate" yaml:"DeadDate,omitempty"`
	PhotoCount    int       `gorm:"default:1" json:"PhotoCount" yaml:"-"`
	PlaceCount    int       `gorm:"default:1" json:"PlaceCount" yaml:"-"`
	CreatedAt     time.Time `json:"CreatedAt" yaml:"-"`
	UpdatedAt     time.Time `json:"UpdatedAt" yaml:"-"`
	DeletedAt     *time.Time  `sql:"index" json:"DeletedAt" yaml:"DeletedAt,omitempty"`
//	PhotoPersonTag []PhotoPersonTag
}


type PersonTag  struct {
	ID            	string    	`gorm:"type:VARBINARY(42);primary_key;auto_increment:false;" json:"PlaceID" yaml:"PlaceID"`
	PersonId     	*string    	`gorm:"type:VARCHAR(42);" json:"PersonId" yaml:"PersonId,omitempty"`
	PhotoUID     	*string    	`gorm:"type:VARCHAR(42);" json:"PhotoUID" yaml:"PhotoUID,omitempty"`
     //MugTag  Information
	PhotoX1    int       `gorm:"default:1" json:"X1" yaml:"X1"`
	PhotoX2    int       `gorm:"default:1" json:"X2" yaml:"X2"`
	PhotoY1    int       `gorm:"default:1" json:"Y1" yaml:"Y1"`
	PhotoY2    int       `gorm:"default:1" json:"Y2" yaml:"Y2"`
	CreatedAt     time.Time `json:"CreatedAt" yaml:"-"`
	UpdatedAt     time.Time `json:"UpdatedAt" yaml:"-"`
	DeletedAt     *time.Time  `sql:"index" json:"DeletedAt" yaml:"DeletedAt,omitempty"`
}
`Is this right for you?

**Comment by lastzero** [comment]
@nicrf Going to look into this in more detail later this month. In principle, our users table is for actual users (or placeholders like anonymous and guest) while we plan to use faces and/or people tables for facial recognition and tagging.

**Comment by MrPeteH** [comment]
@nicrf I agree with you, and @lastzero ... ABSOLUTELY you don't want to limit face recognition to "users."
Remember: faces likely include people who are not even alive anymore... or may be babies :)

I don't know too many one year old coders

**Comment by JAKAMI99** [comment]
fund! to the moon

**Comment by pluja** [comment]
Hey, I'm just curious how is this going? I see on the roadmap that this is in the development state :eyes:

**Comment by lastzero** [comment]
More than happy to start with this. However, these folks are waiting for their PRs to be reviewed and merged: https://github.com/photoprism/photoprism/pulls

Tell me what to do... ignore them?

**Comment by guysoft** [comment]
@lastzero 
I would suggest you [watch this 1 min video of what #foosel says about github issues during this time](https://www.youtube.com/watch?v=AXx6goOcR50&t=592s). People are edgy and in many cases it just happens to be directed at you. I've been seeing it on several issue trackers the past month.
You are doing a wonderful job managing photoprism. This indeed a sort after feature, and it seems the donations are even building up and its indeed on the development roadmap.

**Comment by pluja** [comment]
@lastzero I'm very sorry if it seemed like I was rushing you into this. I didn't mean it that way at all. I was just asking out of curiosity. It is clear that merging PRs is very important!

**Comment by lastzero** [comment]
No worries! We understand everyone is waiting for this... already moved this issue to the development column on our roadmap... might take a couple of weeks until it can be released. We have a zero bug policy, so those (and major UX issues) get fixed first before we continue working on feature requests. Should be in the interest of all users.

**Comment by Iscanniya** [comment]
So nice to hear, you're putting it into development! The acceptance criteria doesn't state the possibility to import allready found faces by google. May this be another rfc? Cause Google Photos puts all the names of found people in the people section of their json-File...

**Comment by lastzero** [comment]
Haven't noticed Google exports this... but also didn't tag people. Can you provide us with an example?

**Comment by Iscanniya** [comment]
Of course, here's one example of one person tagged.

```
{
  "title": "foobar.jpg",
  "description": "",
  "imageViews": "0",
  "creationTime": {
    "timestamp": "1549546798",
    "formatted": "07.02.2019, 13:39:58 UTC"
  },
  "modificationTime": {
    "timestamp": "1612034183",
    "formatted": "30.01.2021, 19:16:23 UTC"
  },
  "photoTakenTime": {
    "timestamp": "1148246652",
    "formatted": "21.05.2006, 21:24:12 UTC"
  },
  "geoData": {
    "latitude": 0.0,
    "longitude": 0.0,
    "altitude": 0.0,
    "latitudeSpan": 0.0,
    "longitudeSpan": 0.0
  },
  "geoDataExif": {
    "latitude": 0.0,
    "longitude": 0.0,
    "altitude": 0.0,
    "latitudeSpan": 0.0,
    "longitudeSpan": 0.0
  },
  "people": [{
    "name": "John Doe"
  }],
  "googlePhotosOrigin": {
    "driveDesktopUploader": {
    }
  }
}
```

**Comment by guysoft** [comment]
That tells you who, but not where.
Not sure how you plan ti define the feature. Ideally face clustering would help you detect new faces. And to cluster you will need a distance metric.

**Comment by el-tiuri** [comment]
If importing previously tagged faces is something you'd want to implement (I think it'd be great!) importing from XMP metadata should probably be doable. It defines an area, coordinates and the name associated with the face. I attached an XMP file (renamed to txt so I could upload it here) with a person tagged. To anonymise it I changed the name to "Dummy data".

[DSCF1911.JPG.txt](https://github.com/photoprism/photoprism/files/5981164/DSCF1911.JPG.txt)

**Comment by lastzero** [comment]
@el-tiuri For this, we need a proper XML / XMP parser first. The one that comes with Go is "not so good" and currently being rewritten from what I know. Once all this work is done, it may be easy to import faces from XMP.

**Comment by pluja** [comment]
I know you want to use a written-in-go library, but maybe this project I just found might help somehow. So I will leave the link to the repo just in case:
https://github.com/exadel-inc/CompreFace

**Comment by lastzero** [comment]
Interesting! So many options... was different when we started... patience pays out :)

https://github.com/esimov/pigo received major improvements as well... I like the simplicity and that it doesn't have external dependencies (as we already have enough to maintain).

**Comment by alexislefebvre** [comment]
Have you considered using a separate Docker image for image detection? Maybe it already exists and it can be plugged through an HTTP API, it would avoid managing the dependencies.

**Comment by lastzero** [comment]
Decided to become @esimov's first GitHub Sponsor :partying_face: :balloon: 

He's maintaining [Pigo](https://github.com/esimov/pigo) and a bunch of other [libraries](https://github.com/esimov?tab=repositories) that look amazing. Can't wait to give them a try!

![yeah](https://user-images.githubusercontent.com/301686/108713877-1ca2de00-7519-11eb-9d5e-80b2e0817036.png)

**Comment by esimov** [comment]
@lastzero thank you, I'm really grateful for your generosity and also for your kind words. 🙏 Also keep in mind that Pigo is still in active development and there is still one big thing left behind in the upcoming future: object recognition and description.

**Comment by lostmaniac** [comment]
How long can I expect to use this feature? Looking forward to face recognition

**Comment by thomaseizinger** [comment]
> How long can I expect to use this feature? Looking forward to face recognition

It is gonna be faster if you join the list of [sponsors](https://github.com/sponsors/photoprism) :)

**Comment by dror3go** [comment]
Is there a plan to support pets recognition, or is it out of the scope of this issue?

**Comment by graciousgrey** [comment]
@dror3go it is out of scope of this issue. But it is something we would love to have as well at some point :)

**Comment by AndreyGurevich** [comment]
Anyway, it will be a great success if the black cat will be recognized as a cat, not as a "dark area of the photo" :)

**Comment by dada051** [comment]
I have already tagged photos (from the Windows Live Photo Gallery) that digiKam reads, so I think it's OK. But I have no results when searching on PhotoPrism. Any idea ?

**Comment by graciousgrey** [comment]
@dada051 PhotoPrism might read other fields than digiKam. You find an overview here: https://docs.photoprism.org/user-guide/advanced/metadata/. Do you know to which metadata field Windows Gallery writes the tags?

**Comment by dada051** [comment]
I'm assuming that digiKam only reads face tag from the standard XPM tag
![image](https://user-images.githubusercontent.com/8973779/117954845-b5c2e300-b317-11eb-9921-ed31ab44ca4f.png)

**Comment by graciousgrey** [comment]
@dada051  If you send us an example xmp file we might be able to add the field

**Comment by dada051** [comment]
Thanks, but I don't have XMP sideloaded. But the document here explain the tags: https://docs.microsoft.com/en-us/windows/win32/wic/-wic-people-tagging

I will try to sideload an XMP to send

**Comment by TYKUHN2** [comment]
I've already implemented go-face natively within Photoprism, unfortunately I hate HTML (and Vue) coding and before I could finish it I got distracted by family and school matters. I suppose I can release a stop-gap measure that is incomplete, but one of the core features (training) would be limited. I also have not tested the implementation yet, because I did not finish the UI, and I did not feel like simulating REST commands at that point.

**Comment by el-tiuri** [comment]
> @dada051 If you send us an example xmp file we might be able to add the field

@graciousgrey I uploaded an XMP file like that on 15 Feb in this thread

**Comment by graciousgrey** [comment]
@el-tiuri Found it. Thank You :)

**Comment by dada051** [comment]
I just found that digiKam uses exiv2 that contains some doc : https://www.exiv2.org/tags-xmp-MP.html

**Comment by dada051** [comment]
Here is an XMP file that contains both "MP:RegionInfo" tag from Windows Gallery and "mwg-rs:Regions" tag that is the standard one I think.

[XMP.txt](https://github.com/photoprism/photoprism/files/6521694/XMP.txt)

**Comment by layanto** [comment]
Any indication on how accurate the face recognition be when available in PhotoPrism? I have been testing a few different face recognition library - MachineBox's FaceBox (https://machinebox.io/docs/facebox), Vladmandic's Face-API (https://github.com/vladmandic/face-api) and Vladmandic's Human (https://github.com/vladmandic/human). FaceBox didn't work very well on my test. Human's face detection was good but face recognition didn't work for me (most likely my error). Face-API worked pretty well but face recognition was nowhere as good as Google Photo's. It is especially bad on face recognition of children. Maybe I need to have a larger set of training photos (in my test, I had about 20-30 training photos per person).

**Comment by lastzero** [comment]
We strive to make it as good as technically possible. How much time we can spend with this depends on our funding, so make sure to become a sponsor if you haven't signed up already.

**Comment by lastzero** [comment]
Development Preview has been updated, see https://docs.photoprism.org/release-notes/.

The new search filter "faces" enables you to exclude faces, or specify a minimum number of faces:
https://docs.photoprism.org/user-guide/organize/search/#search-filter-list

Examples: `faces:no`, `faces:1` is the same as `faces:yes`, `faces:2` finds pictures with at least 2 faces detected.

**Comment by carck** [comment]
Tested the new face detection, it doesn't work very well.  rotated face can not be detected, there is a pending issue in pigo https://github.com/esimov/pigo/issues/9

**Comment by lastzero** [comment]
> Tested the new face detection, it doesn't work very well.

Guess all software projects start like this 😂

**Comment by dada051** [comment]
> Tested the new face detection, it doesn't work very well. rotated face can not be detected, there is a pending issue in pigo [esimov/pigo#9](https://github.com/esimov/pigo/issues/9)

So it works a little. That's a good start :)

**Comment by lastzero** [comment]
Started a new preview build: https://docs.photoprism.app/getting-started/updates/#development-preview

If you're upgrading from a recent preview build, you might have to manually remove the existing `markers_dev` database table before testing. The `_dev` table suffix will be removed for the stable release. It's only temporary to avoid issues when upgrading from a development / preview version to the stable release.

**Comment by geosmart** [comment]
when does face rec feature merge and release?

**Comment by lastzero** [comment]
After we're back from vacation and everyone is happy with the results.

**Comment by rahul003** [comment]
I'm new to the project, but I work on AI and would love to help with this feature. I understand it's already in preview, but I tried switching to preview container and turned on experimental features, I could see that faces: 1 or faces: yes returned some search results. Is there also some support for grouping similar faces together? What's the latest status on development? If there's any discussion somewhere else on the development, and I can pitch in somehow, please let me know.

**Comment by lastzero** [comment]
Hello @rahul003! We're currently trying to grow our team to get hold of all pull and feature requests. That reduces the time we can spend with development short-term. Still busy answering emails (and maintaining our infrastructure) after coming back from vacation.

PhotoPrism currently detects faces and stores face marker positions in the database:

- https://github.com/photoprism/photoprism/blob/develop/internal/face/face.go#L92
- https://github.com/photoprism/photoprism/blob/develop/internal/photoprism/index_mediafile.go#L611

They still need to be normalized (maybe rotated) and clustered. There are different ways to do this, so needs experimenting. You're welcome to take a look, work on a proof-of-concept as unit test, or make suggestions on what libraries to use. When this is done, we can add a people search to the UI.

Note there's an additional issue for using / supporting [DeepStack](https://github.com/photoprism/photoprism/issues/728). However that's an external service / tool from what I know, so we can't compile it in our binary. That's why we're trying to build a native Go solution first. External APIs can be supported later.

**Comment by rahul003** [comment]
Thanks @lastzero , I've made a PR for a POC. Please take a look when you get a chance. I thought it would be good to get some feedback before proceeding further.

**Comment by Extarys** [comment]
Would this allow users to manually tag people on photos where we see faces on the side or from the back that will obviously not be recognized? :smile:

**Comment by lastzero** [comment]
> Would this allow users to manually tag people on photos where we see faces on the side or from the back that will obviously not be recognized?

We'll eventually implement this in our UI. Right now, our development efforts are focused on automatic detection in the backend.

**Comment by lastzero** [comment]
Tweaked face clustering parameters for more accurate results, and added a `photoprism faces reset` command to simplify testing. Development Preview and Demo have been updated: https://demo.photoprism.org/

 Still on our todo: 

- People overview & edit UI 
- Autocomplete for names
- Name / subject search
- XMP subject parsing & import

**Comment by lastzero** [comment]
Major improvement from the previous version. Looks like we need to implement more logic like dynamic thresholds. Goal is a good balance between false positives and negatives, ideally none.

**Comment by totti4ever** [comment]
Wow, can't believe it's happening. soooo looking forward to this.

Excited to see how it works with kids, would be fine in manual corrections were possible

Keep on the great work

**Comment by aravindhsampath** [comment]
Very cool. Looking forward to having my private instance of PhotoPrism with all personal photos, tagging people and being able to simply search for photos of people I know! 

Two things I thought while playing with the demo were :

1. It would be very nice to be able to select multiple photos and simply assign to a label(from a drop-down). At the moment, I can select multiple photos, and when I click edit, I have to add the people names on every single one of them. 

2. Being able to quickly see the "auto-tagged" label or people entry of a picture. Kind of like how Facebook would show tagged people on a photo would be super nice to have. If we could somehow modify the tags while viewing the pictures, without going to another page, it would be even better.

I can create separate issues for them if you think they are worthwhile or align with the thoughts of the devs here.

**Comment by graciousgrey** [comment]
@aravindhsampath 

1. Will be part of #271 
2. You mean changing the face tag in fullscreen mode instead of the edit dialogue?

We plan to support manually tagging faces in a later version. As we need to build a new photo viewer ( #1307 )  I am not sure if it makes sense implementing this before

**Comment by Extarys** [comment]
On the second point, what I understood is that on mouse hover or with a button, boxes with peoples' names should be visible instead of going to the People tab and get the face out of context.

DDG searched for a quick example:
![image](https://user-images.githubusercontent.com/11581433/130164074-bef6f508-b3f1-4bab-aed8-240e8640ff71.png)

**Comment by aravindhsampath** [comment]
@graciousgrey  With #2, I meant to get two things actually. First, being able to see what has been tagged in a picture (items, colours, people etc) so that the user can see what is going on. This may be an additional button in the viewer to "show labels" or something like that. Second, being able to tag people on the picture in view without going to a new tab or page. You already addressed second part by mentioning the building of a new photo viewer, and ability to manually tagging photos coming in a future release. 

So, I am happy to use and view the auto tags for now, and eagerly wait for the ability to add manual tags in future!

**Comment by graciousgrey** [comment]
@aravindhsampath feel free to add a comment to the photo viewer ticket about this :)

**Comment by lastzero** [comment]
Updated demo.photoprism.org with our latest facial recognition improvements. When you correct mistakes with recognizing similar faces, PhotoPrism now remembers this. Simply remove the wrong name by pressing X next to it in the photo edit dialog. Going to work on search & UI next.

**Comment by Extarys** [comment]
**I didn't test this** but I thought about a scenario that might cause some issues.

- If there is a face detected as Sally (cause other photos contain a pre tagged Sally) but it was detected wrong and I change the name to Suzan on that photo, will all the Sally detected faces be renamed to Suzan or just that one picture?

- If so, and I accidentally mistyped Sally's name the first time and I want to rename it, how should I proceed? (Or add her last name later on for example)

**Comment by xaviergxf** [comment]
Is there a way to view all the persons and then when clicking on a person you can see all of their photos?

**Comment by ark-** [comment]
> Is there a way to view all the persons and then when clicking on a person you can see all of their photos?

My understanding is that this feature is currently still being worked on. 

@lastzero A quick win in this regard could be adding the person name to a photos tags? And so the existing search system will pick up the person name.

**Comment by lastzero** [comment]
> * If there is a face detected as Sally (cause other photos contain a pre tagged Sally) but it was detected wrong and I change the name to Suzan on that photo, will all the Sally detected faces be renamed to Suzan or just that one picture?
> * If so, and I accidentally mistyped Sally's name the first time and I want to rename it, how should I proceed? (Or add her last name later on for example)

If it's the right person and you just want to change the name (for all matched faces), you simply change the name. That's it.

If the wrong person was recognized:

- Click on the X next to the name, this will tell PhotoPrism it is a different person. Then enter the actual name if you know it.
- Soon the field has autocomplete, so you don't need to remember the exact names as it is currently the case.

**Comment by lastzero** [comment]
> My understanding is that this feature is currently still being worked on.

Exactly, we're doing the backend first before we finish the UI. It's a lot of math and testing involved. We're testing with more and more pictures until all obvious issues are fixed.
 
> @lastzero A quick win in this regard could be adding the person name to a photos tags? And so the existing search system will pick up the person name.

We'd still need to keep both in sync as users would report this as bug otherwise. So it's better to implement a subject search from the start - also since being able to search for one or more people is a core feature.

Developing the UI and search is really not that much work compared to the backend. There's only so much work we can get done at a time though. We're already working every weekend, and also need time for taking care of support requests every day.

**Comment by lastzero** [comment]
Demo has been updated with our latest improvements, you can now search for one or more subjects: https://demo.photoprism.org/browse?view=cards&order=added&q=people%3A%22Otto%20%26%20Monika%22

**Comment by lastzero** [comment]
Started a new [Development Preview](https://docs.photoprism.org/release-notes/) / [Demo](https://demo.photoprism.org/) [build](https://drone.photoprism.app/photoprism/photoprism/1713): 
- You can soon search for one or more people without explicitly using the `people:...` / `subjects:...` search filter!
- Subject names can be combined with label names, or keywords.
- Album and label thumbnails should load faster as the API now directly returns preview image hashes.

**Comment by lastzero** [comment]
As always, you'll have to reindex for testing the latest improvements at home (or use our [public demo](https://demo.photoprism.org/browse)):

- Face crops (thumbnails) are now generated on the server. The API supports 160px and 320px square tiles. Quality may be better, performance too (depending on your device and network connection).
- Autocomplete sorting when adding names to faces has been improved.
- Search is a lot more powerful now as well. For example, you can search for "John Doe at the Botanical Garden" or "Obama in Berlin", which will find all pictures of the Obama family in Berlin. You can also search for "Michael & Theresa with sunglasses" to find pictures of Michael AND Theresa where at least on person is wearing sunglasses.

People overview / edit page will be implemented when we're back in about 2 weeks. Enjoy! :sunglasses:

**Comment by lastzero** [comment]
Most common use cases should be covered with this. We've spent the day tweaking detection and clustering parameters for best results. Looks good for now. Going to test with more pictures tomorrow.

There are a few edge cases the initial release won't handle, like when you move, distort, or otherwise change faces in already indexed originals. PhotoPrism doesn't track such image content changes and therefore won't update people matches / face thumbnails along with it (it may detect new faces, which can lead to duplicates e.g. when the same face is now at a different position). It's generally best to save JPEG edits as a new file as it degrades image quality as well. We can handle this and other edge cases later, should they become an issue in practice.

We also expect less accurate results if you frequently change to which person a face belongs as this breaks up clusters (of similar faces) and creates new clusters based on the new information. Keep this in mind when testing.

Our new Docker base image is still uploading to Docker Hub. Once that is done, we'll start a [Development Preview](https://docs.photoprism.org/release-notes/) build as release candidate. Feedback welcome!

**Comment by lastzero** [comment]
Manual face tagging and XMP import will be part of a later release, @graciousgrey will create issues for this 👍

**Comment by lastzero** [comment]
Tests with large libraries show issues with very similar faces, like sisters. Next build will be more strict. You'll have to name more faces instead.

**Comment by mike-lloyd03** [comment]
It seems to be having trouble finding faces if the person is wearing a hat and sunglasses. Is there anything that can be done to improve this?

**Comment by Jeongwoo55** [comment]
It also seems to have trouble finding faces which are slightly cutoff. Roughly 10% of the face (mostly hair) is cutoff and all other features of their face is clearly visible yet they are undetected. Not sure if much can be done to improve this.

**Comment by lastzero** [comment]
Happy to take a look at edge cases after our release. It's possible to find more faces, the question is how much time and money we can throw at it. As always, you're welcome to send samples for testing! 🕵️‍♀️

**Comment by mike-lloyd03** [comment]
I'd be happy to. Where do I send them?

I imagine once manual face tagging gets implemented, these problems will likely solve themselves.

**Comment by lastzero** [comment]
While finding cut off faces should be possible, it makes sense those can't be properly compared or clustered as information is missing. As mentioned in the issue title, the goal was recognizing faces not just finding them. You're welcome to send files (or links to files) to hello@photoprism.app. Thank you! 👍

**Comment by lastzero** [comment]
[Development Preview](https://docs.photoprism.org/release-notes/) and [demo.photoprism.org](https://demo.photoprism.org/) have been updated for testing 🌈

**Comment by jeremywillans** [comment]
@lastzero, awesome work with all your efforts implementing this!

I am testing the latest develop build and in the new Faces menu, it does not appear to list all the unrecognized faces. 
In my sample batch of photos, there are a few different faces and only two appear in that list, but if I edit one of the other photos, there are other new faces which I can tag with names. Once tagged, they are correctly added to the recognized list. 

Not sure if this is a known issue?

**Comment by nagarrido** [comment]
@jeremywillans same issue here! Awesome work by the way, I've been waiting a long time for this!

**Comment by graciousgrey** [comment]
@jeremywillans @nagarrido This is expected in the *New* section only face clusters show up (minimum 4 faces). Single faces that are not clustered are only visible in the edit dialogue for now.

**Comment by lastzero** [comment]
@jeremywillans @nagarrido We've just improved the existing "face" search filter so that you can more easily find pictures with unclustered faces (meaning less than 4 samples exist):

```
face:new
```

We could also show every single unclustered face on the "new faces" page. Keep in mind it can be thousands e.g. from shampoo bottles and screenshots. Not sure if it would add real value or just noise.

**Comment by nagarrido** [comment]
@lastzero great! I agree, maybe showing ALL unclustered faces in the new faces section would not be a good idea. The search item seems like the best idea. Again, thanks for everything!

**Comment by tribut** [comment]
Maybe there could be a link to this search at the bottom ("show all new faces" or similar)? Otherwise this cutoff seems confusing without explanation.

**Comment by graciousgrey** [comment]
@tribut great idea :)

**Comment by lastzero** [comment]
Started a new build with this. Let us know if it works for you!

https://docs.photoprism.app/getting-started/updates/#development-preview

**Comment by nagarrido** [comment]
@lastzero it's working perfecty! I've been using a test folder for face recognition, so I will reindex overnight to see how it works with my whole library. I think this solution will satisfy all users, as it still keeps the "people" section organized and still lets us (the picky ones) with a great way to tag every face. Thanks for the great work!

**Comment by ThomasF34** [comment]
@lastzero I just tested it out it's really great ! Thanks for everything 😍 
Little question though: When we use this `Show all new faces` we arrive on the search page and from here it's quite difficult to review the faces (add name or indicate that's it's not a face, as we can do on People > New page)
Do you thing it would be feasible to improve the review process by seeing all the new faces with the ability to act on it ?

**Comment by graciousgrey** [comment]
@ThomasF34 I created a follow up ticket for this: https://github.com/photoprism/photoprism/issues/1556

**Comment by lastzero** [comment]
Note you can easily reset and re-index faces if needed, for example after upgrading or when you've made a mistake that takes long to fix otherwise:
```
docker-compose exec photoprism photoprism faces reset -f
docker-compose exec photoprism photoprism faces index
```
Naming usually just takes a couple of minutes. With our new version released yesterday, you may even get better detection results with fewer false positives. Instead of executing the commands with Docker Compose, you may skip "docker-compose exec photoprism" and directly run the "photoprism faces ..." commands in a terminal wherever PhotoPrism is installed.

*If you're happy with the results and didn't make a mistake you want to fix anyway, you don't have to reset existing people & faces.* The new release is better at ignoring false positives, which leads to less noise and better performance (as less faces need to compared with each other). Also it enables you to detect smaller faces if you want to (see `PHOTOPRISM_FACE_SIZE` in our [docs](https://docs.photoprism.org/getting-started/config-options/)).

**Comment by simoncoul** [comment]
@lastzero would it be possible to get more documentation of the facial recognition parameters? Currently it does not indicate how the parameters influence the recognition.

**Comment by lastzero** [comment]
`PHOTOPRISM_FACE_SCORE` determines the quality of the face detection, lower quality means faces are less likely to be a face. You may experiment with this, we think the default is the best setting.

`PHOTOPRISM_FACE_SIZE` determines the minimum size of faces in pixels. You may lower this to 40 or 30 to detect smaller faces. Such small faces won't be used for clustering though, but they can be recognized based on existing clusters. Detecting more faces will reduce performance. You may experiment with this like we did for many weeks.

`PHOTOPRISM_FACE_OVERLAP` determines how much detected faces may overlap. You may experiment with this, we think the default is the best setting.

`PHOTOPRISM_FACE_CLUSTER_CORE` determines how many faces are needed to form a cluster core for recognizing people from individual faces. You may experiment with this, we think the default is the best setting.

`PHOTOPRISM_FACE_CLUSTER_DIST` and `PHOTOPRISM_FACE_MATCH_DIST` influence clustering and detection. See https://towardsdatascience.com/machine-learning-clustering-dbscan-determine-the-optimal-value-for-epsilon-eps-python-example-3100091cfbc to learn more. If you're not familiar with machine learning and math, you better don't play with it.

**Comment by lastzero** [comment]
You may also want to read this to get a better understanding how clustering works (explaining this in detail is out of scope for our docs, especially since we need to continue working on multi-user support now and don't have a lot of time): https://en.wikipedia.org/wiki/DBSCAN

**Comment by alexislefebvre** [comment]
> `PHOTOPRISM_FACE_SCORE` determines the quality of the face detection, lower quality means faces are less likely to be a face. You may experiment with this, we think the default is the best setting.

The default value is 9, is it 9/9, 9%? I suggest we add the minimal and maximal values for `PHOTOPRISM_FACE_SCORE` to the doc:

https://github.com/photoprism/photoprism-docs/blob/0bba8a88d0668ed0970dad466a21692952d10e44/docs/getting-started/config-options.md?plain=1#L81

**Comment by lastzero** [comment]
I've spent a couple of hours yesterday improving our command and config flag descriptions (and also added two more options for you to play with, if you feel like this can lead to better results for your use case):
```
   --face-size PIXELS                       min face size in PIXELS (default: 40) [$PHOTOPRISM_FACE_SIZE]
   --face-score THRESHOLD                   quality THRESHOLD for faces (default: 9) [$PHOTOPRISM_FACE_SCORE]
   --face-overlap PERCENT                   face area overlap threshold in PERCENT (default: 42) [$PHOTOPRISM_FACE_OVERLAP]
   --face-cluster-size PIXELS               min size of faces forming a cluster in PIXELS (default: 80) [$PHOTOPRISM_FACE_CLUSTER_SIZE]
   --face-cluster-score THRESHOLD           quality THRESHOLD for faces forming a cluster (default: 15) [$PHOTOPRISM_FACE_CLUSTER_SCORE]
   --face-cluster-core NUMBER               NUMBER of faces forming a cluster core (default: 4) [$PHOTOPRISM_FACE_CLUSTER_CORE]
   --face-cluster-dist DISTANCE             similarity DISTANCE of faces forming a cluster core (default: 0.64) [$PHOTOPRISM_FACE_CLUSTER_DIST]
   --face-match-dist OFFSET                 similarity OFFSET for matching faces with existing clusters (default: 0.46) [$PHOTOPRISM_FACE_MATCH_DIST]
```

@alexislefebvre The face detection library is https://github.com/esimov/pigo, maybe you find more information there in the docs or code. I guess the score can go from 1 to 1000? No idea really, as those don't play a role in practice. That's either way too high or too low for good results. So you shouldn't use these values anyway. We've tested with many different pictures and decided to go for a min score of 9, which is dynamically increased for smaller faces. Machine learning often is more art than science.

Especially the similarity distances ("epsilon") for the face embedding vectors are super sensitive, so any significantly larger or smaller values won't work in practice. More information:

- https://dl.photoprism.org/research/A%20Comparison%20of%20Clustering%20Algorithms%20for%20Faces.pdf
- https://towardsdatascience.com/how-dbscan-works-and-why-should-i-use-it-443b4a191c80
- https://scikit-learn.org/stable/auto_examples/cluster/plot_dbscan.html

**Comment by lastzero** [comment]
Continued revising configuration flag and command descriptions today, docs have been updated (noticed they contained a few obvious mistakes like a wrong default value for FFmpeg capture buffers, > 500 instead of 32):
https://docs.photoprism.org/getting-started/config-options/

**Comment by alexislefebvre** [comment]
> @alexislefebvre The face detection library is https://github.com/esimov/pigo, maybe you find more information there in the docs or code. I guess the score can go from 1 to 1000? No idea really, as those don't play a role in practice. That's either way too high or too low for good results. So you shouldn't use these values anyway. We've tested with many different pictures and decided to go for a min score of 9, which is dynamically increased for smaller faces. Machine learning often is more art than science.

The author of this library [answered to my question](https://github.com/esimov/pigo/issues/50#issuecomment-938430620), I didn't understood everything but you may find it interesting.

**Comment by lastzero** [comment]
Thanks! In fact we've recently changed IoU in the stable version and it seems to have slightly improved results.

**Comment by grahamPegNetwork** [comment]
I just saw the comments here, way to go team! Docker pull, restart and look at it go! There has been a lot of activity since I last checked. Ultimately, who is most responsible for this feature happening? Is this the work of @lastzero ?
